File: nodejs/src/api/providers/anthropic.ts
/**
 * Anthropic (Claude) provider adapter.
 *
 * Converts between OpenAI and Anthropic API formats.
 */

import { v4 as uuidv4 } from "uuid";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface ProviderConfig {
  baseUrl?: string;
  apiKey: string;
  extraHeaders?: Record<string, string>;
  timeout?: number;
}

export interface ChatCompletionRequest {
  model: string;
  messages: Array<{
    role: string;
    content: string | any[];
  }>;
  stream?: boolean;
  temperature?: number;
  top_p?: number;
  max_tokens?: number;
  stop?: string | string[];
  [key: string]: unknown;
}

export interface Usage {
  input_tokens: number;
  output_tokens: number;
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const DEFAULT_TIMEOUT = 120_000;
const MAX_RETRIES = 2;
const RETRY_DELAY = 500;
const ANTHROPIC_VERSION = "2023-06-01";

const MODEL_MAP: Record<string, string> = {
  "claude-3-opus": "claude-3-opus-20240229",
  "claude-3-sonnet": "claude-3-sonnet-20240229",
  "claude-3-haiku": "claude-3-haiku-20240307",
  "claude-3.5-sonnet": "claude-3-5-sonnet-20241022",
  "claude-3.5-haiku": "claude-3-5-haiku-20241022",
  "claude-4-opus": "claude-opus-4-20250514",
  "claude-4-sonnet": "claude-sonnet-4-20250514",
};

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export async function chatCompletion(
  requestData: ChatCompletionRequest,
  providerConfig: ProviderConfig
): Promise<Record<string, unknown> | AsyncGenerator<Uint8Array>> {
  const baseUrl = (providerConfig.baseUrl ?? "https://api.anthropic.com").replace(
    /\/+$/,
    ""
  );
  const apiKey = providerConfig.apiKey;
  const timeout = providerConfig.timeout ?? DEFAULT_TIMEOUT;
  const extraHeaders = providerConfig.extraHeaders ?? {};
  const isStream = requestData.stream === true;

  const anthropicBody = toAnthropicRequest(requestData);
  const url = `${baseUrl}/v1/messages`;

  const headers: Record<string, string> = {
    "x-api-key": apiKey,
    "anthropic-version": ANTHROPIC_VERSION,
    "Content-Type": "application/json",
    ...extraHeaders,
  };

  return doRequest(url, headers, anthropicBody, timeout, isStream, requestData.model);
}

// ---------------------------------------------------------------------------
// Format conversion: OpenAI -> Anthropic
// ---------------------------------------------------------------------------

function toAnthropicRequest(openaiReq: ChatCompletionRequest): Record<string, any> {
  const model = openaiReq.model ?? "claude-3-sonnet";
  const mappedModel = MODEL_MAP[model] ?? model;

  const messagesRaw = openaiReq.messages ?? [];

  let systemText = "";
  const anthropicMessages: Array<{ role: string; content: any }> = [];

  for (const msg of messagesRaw) {
    const role = msg.role ?? "user";
    const content = msg.content;

    if (role === "system") {
      if (Array.isArray(content)) {
        const parts = content
          .filter((p: any) => p.type === "text")
          .map((p: any) => p.text ?? "");
        systemText += parts.join("\n");
      } else {
        systemText += content + "\n";
      }
    } else {
      const anthropicContent = convertContent(content);
      anthropicMessages.push({ role, content: anthropicContent });
    }
  }

  // Merge consecutive same-role messages (Anthropic requirement)
  const merged: Array<{ role: string; content: any }> = [];
  for (const msg of anthropicMessages) {
    if (merged.length > 0 && merged[merged.length - 1].role === msg.role) {
      const prev = merged[merged.length - 1].content;
      const curr = msg.content;
      merged[merged.length - 1].content =
        (typeof prev === "string" ? prev : String(prev)) +
        "\n" +
        (typeof curr === "string" ? curr : String(curr));
    } else {
      merged.push({ role: msg.role, content: msg.content });
    }
  }

  const body: Record<string, any> = {
    model: mappedModel,
    messages: merged,
    max_tokens: openaiReq.max_tokens ?? 4096,
  };

  if (systemText.trim()) {
    body.system = systemText.trim();
  }

  if (openaiReq.temperature !== undefined) body.temperature = openaiReq.temperature;
  if (openaiReq.top_p !== undefined) body.top_p = openaiReq.top_p;
  if (openaiReq.stop !== undefined) {
    body.stop_sequences = Array.isArray(openaiReq.stop)
      ? openaiReq.stop
      : [openaiReq.stop];
  }
  if (openaiReq.stream !== undefined) body.stream = openaiReq.stream;

  return body;
}

function convertContent(content: any): any {
  if (typeof content === "string") return content;
  if (Array.isArray(content)) {
    const parts: Array<Record<string, any>> = [];
    for (const part of content) {
      const ptype = part.type ?? "text";
      if (ptype === "text") {
        parts.push({ type: "text", text: part.text ?? "" });
      } else if (ptype === "image_url") {
        const url: string = part.image_url?.url ?? "";
        if (url.startsWith("data:")) {
          const mediaType = url.includes(":") ? url.split(";")[0].split(":")[1] : "image/png";
          const b64data = url.includes(",") ? url.split(",", 2)[1] : "";
          parts.push({
            type: "image",
            source: { type: "base64", media_type: mediaType, data: b64data },
          });
        }
      }
    }
    return parts.length > 0 ? parts : String(content);
  }
  return String(content);
}

// ---------------------------------------------------------------------------
// Format conversion: Anthropic -> OpenAI
// ---------------------------------------------------------------------------

function toOpenAIResponse(
  anthropicResp: Record<string, any>,
  originalModel: string
): Record<string, any> {
  const contentBlocks: Array<Record<string, any>> = anthropicResp.content ?? [];
  const textParts: string[] = [];
  for (const block of contentBlocks) {
    if (block.type === "text") textParts.push(block.text ?? "");
  }
  const text = textParts.join("");

  const usageIn = anthropicResp.usage ?? {};
  const inputTokens = usageIn.input_tokens ?? 0;
  const outputTokens = usageIn.output_tokens ?? 0;

  const stopReason = anthropicResp.stop_reason ?? "end_turn";
  const finishReason = mapStopReason(stopReason);

  return {
    id: `chatcmpl-${uuidv4().replace(/-/g, "").slice(0, 24)}`,
    object: "chat.completion",
    created: Math.floor(Date.now() / 1000),
    model: originalModel,
    choices: [
      {
        index: 0,
        message: { role: "assistant", content: text },
        finish_reason: finishReason,
      },
    ],
    usage: {
      prompt_tokens: inputTokens,
      completion_tokens: outputTokens,
      total_tokens: inputTokens + outputTokens,
    },
  };
}

function mapStopReason(reason: string): string {
  const mapping: Record<string, string> = {
    end_turn: "stop",
    max_tokens: "length",
    stop_sequence: "stop",
  };
  return mapping[reason] ?? "stop";
}

// ---------------------------------------------------------------------------
// HTTP request with retry
// ---------------------------------------------------------------------------

async function doRequest(
  url: string,
  headers: Record<string, string>,
  body: Record<string, any>,
  timeout: number,
  isStream: boolean,
  originalModel: string
): Promise<Record<string, unknown> | AsyncGenerator<Uint8Array>> {
  let lastError: Error | null = null;

  for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
    try {
      if (isStream) {
        return streamResponse(url, headers, body, timeout, originalModel);
      }

      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), timeout);

      const resp = await fetch(url, {
        method: "POST",
        headers,
        body: JSON.stringify(body),
        signal: controller.signal,
      });
      clearTimeout(timer);

      if (!resp.ok) {
        const errorBody = await resp.text();
        const error = new Error(`Anthropic API error ${resp.status}: ${errorBody}`);
        (error as any).status = resp.status;
        throw error;
      }

      const data = await resp.json();
      return toOpenAIResponse(data, originalModel);
    } catch (err: any) {
      lastError = err;
      const status: number | undefined = err.status;

      if (status && status >= 400 && status < 500 && status !== 429) {
        throw wrapError(err);
      }

      if (attempt < MAX_RETRIES) {
        console.warn(
          `Anthropic request failed (attempt ${attempt + 1}/${MAX_RETRIES + 1}): ${err} – retrying`
        );
        await sleep(RETRY_DELAY * (attempt + 1));
        continue;
      }

      throw wrapError(err);
    }
  }

  throw wrapError(lastError!);
}

// ---------------------------------------------------------------------------
// Streaming: Anthropic SSE -> OpenAI SSE
// ---------------------------------------------------------------------------

async function* streamResponse(
  url: string,
  headers: Record<string, string>,
  body: Record<string, any>,
  timeout: number,
  originalModel: string
): AsyncGenerator<Uint8Array> {
  const completionId = `chatcmpl-${uuidv4().replace(/-/g, "").slice(0, 24)}`;
  const created = Math.floor(Date.now() / 1000);
  const encoder = new TextEncoder();

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeout);

  const resp = await fetch(url, {
    method: "POST",
    headers,
    body: JSON.stringify(body),
    signal: controller.signal,
  });
  clearTimeout(timer);

  if (!resp.ok) {
    const errorBody = await resp.text();
    throw new Error(`Anthropic API error ${resp.status}: ${errorBody}`);
  }

  const reader = resp.body?.getReader();
  if (!reader) throw new Error("No response body for streaming");

  const decoder = new TextDecoder();
  let buffer = "";

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() ?? "";

      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed.startsWith("data: ")) continue;
        const payload = trimmed.slice(6).trim();
        if (!payload) continue;

        let event: Record<string, any>;
        try {
          event = JSON.parse(payload);
        } catch {
          continue;
        }

        const eventType = event.type ?? "";

        if (eventType === "content_block_delta") {
          const deltaText = event.delta?.text ?? "";
          const chunk = buildStreamChunk(completionId, created, originalModel, deltaText);
          yield encoder.encode(`data: ${JSON.stringify(chunk)}\n\n`);
        } else if (eventType === "message_stop") {
          const chunk = buildStreamChunk(
            completionId, created, originalModel, "", "stop"
          );
          yield encoder.encode(`data: ${JSON.stringify(chunk)}\n\n`);
          yield encoder.encode("data: [DONE]\n\n");
          return;
        } else if (eventType === "message_delta") {
          const stopReason = event.delta?.stop_reason ?? "end_turn";
          const finish = mapStopReason(stopReason);
          const usage = event.usage ?? {};
          const chunk = buildStreamChunk(
            completionId, created, originalModel, "", finish, usage
          );
          yield encoder.encode(`data: ${JSON.stringify(chunk)}\n\n`);
        } else if (eventType === "error") {
          const err = event.error ?? {};
          throw new Error(
            `Anthropic stream error: ${err.type ?? "unknown"}: ${err.message ?? "unknown error"}`
          );
        }
      }
    }
  } finally {
    reader.releaseLock();
  }

  yield encoder.encode("data: [DONE]\n\n");
}

function buildStreamChunk(
  completionId: string,
  created: number,
  model: string,
  text: string,
  finishReason: string | null = null,
  usage?: Record<string, any>
): Record<string, any> {
  const chunk: Record<string, any> = {
    id: completionId,
    object: "chat.completion.chunk",
    created,
    model,
    choices: [
      {
        index: 0,
        delta: text ? { content: text } : {},
        finish_reason: finishReason,
      },
    ],
  };

  if (usage) {
    chunk.usage = {
      prompt_tokens: usage.input_tokens ?? 0,
      completion_tokens: usage.output_tokens ?? 0,
      total_tokens: (usage.input_tokens ?? 0) + (usage.output_tokens ?? 0),
    };
  }

  return chunk;
}

// ---------------------------------------------------------------------------
// Usage extraction
// ---------------------------------------------------------------------------

export function extractUsage(responseData: Record<string, any>): Usage {
  const usage = responseData.usage ?? {};
  return {
    input_tokens: usage.prompt_tokens ?? 0,
    output_tokens: usage.completion_tokens ?? 0,
  };
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function wrapError(err: Error): Error {
  if (err.name === "AbortError") {
    return new Error("Anthropic API request timed out");
  }
  return err;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: python/api/providers/anthropic.py
"""
Anthropic (Claude) provider adapter.

Converts between OpenAI and Anthropic API formats.
"""

from __future__ import annotations

import json
import logging
import asyncio
import time
import uuid
from typing import Any, AsyncIterator

import httpx

from api.thinking_parser import inject_for_anthropic

logger = logging.getLogger(__name__)

DEFAULT_TIMEOUT = 120.0
MAX_RETRIES = 2
RETRY_DELAY = 0.5

ANTHROPIC_VERSION = "2023-06-01"

# Map common OpenAI model names to Anthropic model IDs.
MODEL_MAP: dict[str, str] = {
    "claude-3-opus": "claude-3-opus-20240229",
    "claude-3-sonnet": "claude-3-sonnet-20240229",
    "claude-3-haiku": "claude-3-haiku-20240307",
    "claude-3.5-sonnet": "claude-3-5-sonnet-20241022",
    "claude-3.5-haiku": "claude-3-5-haiku-20241022",
    "claude-4-opus": "claude-opus-4-20250514",
    "claude-4-sonnet": "claude-sonnet-4-20250514",
}


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

async def chat_completion(
    request_data: dict[str, Any],
    provider_config: dict[str, Any],
) -> dict[str, Any] | AsyncIterator[bytes]:
    """Send a chat completion request to Anthropic.

    Parameters
    ----------
    request_data:
        OpenAI-format request body.
    provider_config:
        {
            "base_url": "https://api.anthropic.com",  # optional override
            "api_key": "sk-ant-...",
            "extra_headers": {},
            "timeout": 120,
        }
    """
    base_url = provider_config.get("base_url", "https://api.anthropic.com").rstrip("/")
    api_key = provider_config["api_key"]
    timeout = provider_config.get("timeout", DEFAULT_TIMEOUT)
    extra_headers = provider_config.get("extra_headers", {})
    is_stream = request_data.get("stream", False)

    anthropic_body = _to_anthropic_request(request_data)
    url = f"{base_url}/v1/messages"

    headers = {
        "x-api-key": api_key,
        "anthropic-version": ANTHROPIC_VERSION,
        "content-type": "application/json",
        **extra_headers,
    }

    return await _do_request(url, headers, anthropic_body, timeout, is_stream, request_data)


# ---------------------------------------------------------------------------
# Format conversion: OpenAI -> Anthropic
# ---------------------------------------------------------------------------

def _to_anthropic_request(openai_req: dict[str, Any]) -> dict[str, Any]:
    """Convert an OpenAI chat completion request to Anthropic format."""

    model = openai_req.get("model", "claude-3-sonnet")
    mapped_model = MODEL_MAP.get(model, model)

    messages_raw: list[dict[str, Any]] = openai_req.get("messages", [])

    # Separate system message from the rest.
    system_text = ""
    anthropic_messages: list[dict[str, Any]] = []

    for msg in messages_raw:
        role = msg.get("role", "user")
        content = msg.get("content", "")

        if role == "system":
            # Accumulate system messages.
            if isinstance(content, list):
                parts = [p.get("text", "") for p in content if p.get("type") == "text"]
                system_text += "\n".join(parts)
            else:
                system_text += content + "\n"
        else:
            # Anthropic uses "user" and "assistant" roles.
            if role == "tool":
                # Convert tool messages to user role for simplicity.
                anthropic_messages.append({"role": "user", "content": str(content)})
            else:
                anthropic_content = _convert_content(content)
                anthropic_messages.append({
                    "role": role,
                    "content": anthropic_content,
                })

    # Merge consecutive same-role messages (Anthropic requirement).
    merged: list[dict[str, Any]] = []
    for msg in anthropic_messages:
        if merged and merged[-1]["role"] == msg["role"]:
            # Concatenate content
            prev = merged[-1]["content"]
            curr = msg["content"]
            if isinstance(prev, str) and isinstance(curr, str):
                merged[-1]["content"] = prev + "\n" + curr
            else:
                merged[-1]["content"] = str(prev) + "\n" + str(curr)
        else:
            merged.append(msg)

    body: dict[str, Any] = {
        "model": mapped_model,
        "messages": merged,
        "max_tokens": openai_req.get("max_tokens", 4096),
    }

    if system_text.strip():
        body["system"] = system_text.strip()

    # Optional parameters
    if "temperature" in openai_req:
        body["temperature"] = openai_req["temperature"]
    if "top_p" in openai_req:
        body["top_p"] = openai_req["top_p"]
    if "stop" in openai_req:
        stop = openai_req["stop"]
        if isinstance(stop, str):
            stop = [stop]
        body["stop_sequences"] = stop
    if "stream" in openai_req:
        body["stream"] = openai_req["stream"]

    # Inject thinking params if thinking_effort is set
    if openai_req.get("thinking_effort") is not None:
        inject_for_anthropic(body, openai_req["thinking_effort"])

    return body


def _convert_content(content: Any) -> Any:
    """Convert OpenAI content to Anthropic content format."""
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        # OpenAI multi-part content
        parts: list[dict[str, Any]] = []
        for part in content:
            ptype = part.get("type", "text")
            if ptype == "text":
                parts.append({"type": "text", "text": part.get("text", "")})
            elif ptype == "image_url":
                url = part.get("image_url", {}).get("url", "")
                if url.startswith("data:"):
                    # base64 image
                    media_type = url.split(";")[0].split(":")[1] if ":" in url else "image/png"
                    b64data = url.split(",", 1)[1] if "," in url else ""
                    parts.append({
                        "type": "image",
                        "source": {
                            "type": "base64",
                            "media_type": media_type,
                            "data": b64data,
                        },
                    })
        return parts if parts else str(content)
    return str(content)


# ---------------------------------------------------------------------------
# Format conversion: Anthropic -> OpenAI
# ---------------------------------------------------------------------------

def _to_openai_response(
    anthropic_resp: dict[str, Any],
    original_model: str,
) -> dict[str, Any]:
    """Convert an Anthropic response to OpenAI format."""

    content_blocks = anthropic_resp.get("content", [])
    text_parts: list[str] = []
    thinking_parts: list[str] = []
    for block in content_blocks:
        if block.get("type") == "text":
            text_parts.append(block.get("text", ""))
        elif block.get("type") == "thinking":
            thinking_parts.append(block.get("thinking", ""))

    text = "".join(text_parts)
    reasoning_content = "".join(thinking_parts) if thinking_parts else None

    usage_in = anthropic_resp.get("usage", {})
    input_tokens = usage_in.get("input_tokens", 0)
    output_tokens = usage_in.get("output_tokens", 0)

    stop_reason = anthropic_resp.get("stop_reason", "end_turn")
    finish_reason = _map_stop_reason(stop_reason)

    message: dict[str, Any] = {
        "role": "assistant",
        "content": text,
    }
    if reasoning_content is not None:
        message["reasoning_content"] = reasoning_content

    return {
        "id": f"chatcmpl-{uuid.uuid4().hex[:24]}",
        "object": "chat.completion",
        "created": int(time.time()),
        "model": original_model,
        "choices": [
            {
                "index": 0,
                "message": message,
                "finish_reason": finish_reason,
            }
        ],
        "usage": {
            "prompt_tokens": input_tokens,
            "completion_tokens": output_tokens,
            "total_tokens": input_tokens + output_tokens,
        },
    }


def _map_stop_reason(reason: str) -> str:
    mapping = {
        "end_turn": "stop",
        "max_tokens": "length",
        "stop_sequence": "stop",
    }
    return mapping.get(reason, "stop")


# ---------------------------------------------------------------------------
# HTTP request with retry
# ---------------------------------------------------------------------------

async def _do_request(
    url: str,
    headers: dict[str, str],
    body: dict[str, Any],
    timeout: float,
    is_stream: bool,
    original_request: dict[str, Any],
) -> dict[str, Any] | AsyncIterator[bytes]:

    original_model = original_request.get("model", body.get("model", ""))
    last_exc: Exception | None = None

    for attempt in range(MAX_RETRIES + 1):
        try:
            if is_stream:
                return _stream_response(url, headers, body, timeout, original_model)

            async with httpx.AsyncClient(timeout=timeout) as client:
                resp = await client.post(url, headers=headers, json=body)
                resp.raise_for_status()
                data = resp.json()
                return _to_openai_response(data, original_model)

        except (httpx.HTTPStatusError, httpx.ConnectError, httpx.ReadTimeout) as exc:
            last_exc = exc
            status = getattr(getattr(exc, "response", None), "status_code", None)

            if status is not None and 400 <= status < 500 and status != 429:
                raise _wrap_error(exc) from exc

            if attempt < MAX_RETRIES:
                logger.warning(
                    "Anthropic request failed (attempt %d/%d): %s – retrying",
                    attempt + 1,
                    MAX_RETRIES + 1,
                    exc,
                )
                await asyncio.sleep(RETRY_DELAY * (attempt + 1))
                continue

            raise _wrap_error(exc) from exc

    raise _wrap_error(last_exc) from last_exc  # type: ignore[arg-type]


async def _stream_response(
    url: str,
    headers: dict[str, str],
    body: dict[str, Any],
    timeout: float,
    original_model: str,
) -> AsyncIterator[bytes]:
    """Convert Anthropic SSE stream to OpenAI SSE stream."""

    completion_id = f"chatcmpl-{uuid.uuid4().hex[:24]}"
    created = int(time.time())

    async with httpx.AsyncClient(timeout=timeout) as client:
        async with client.stream("POST", url, headers=headers, json=body) as resp:
            resp.raise_for_status()

            async for line in resp.aiter_lines():
                if not line.startswith("data: "):
                    continue
                payload = line[len("data: "):]
                if not payload.strip():
                    continue

                try:
                    event = json.loads(payload)
                except json.JSONDecodeError:
                    continue

                event_type = event.get("type", "")

                if event_type == "content_block_delta":
                    delta = event.get("delta", {})
                    delta_type = delta.get("type", "")
                    if delta_type == "thinking_delta":
                        # Anthropic thinking content: delta.thinking
                        thinking_text = delta.get("thinking", "")
                        if thinking_text:
                            chunk = _build_stream_chunk(
                                completion_id, created, original_model, "",
                                reasoning_content=thinking_text,
                            )
                            yield f"data: {json.dumps(chunk)}\n\n".encode("utf-8")
                    elif delta_type == "signature_delta":
                        # Signature is a security token, no need to forward
                        pass
                    else:
                        # text_delta or other: delta.text
                        delta_text = delta.get("text", "")
                        if delta_text:
                            chunk = _build_stream_chunk(
                                completion_id, created, original_model, delta_text
                            )
                            yield f"data: {json.dumps(chunk)}\n\n".encode("utf-8")

                elif event_type == "message_stop":
                    # Send final chunk with finish_reason
                    chunk = _build_stream_chunk(
                        completion_id, created, original_model, "",
                        finish_reason="stop",
                    )
                    yield f"data: {json.dumps(chunk)}\n\n".encode("utf-8")
                    yield b"data: [DONE]\n\n"
                    return

                elif event_type == "message_delta":
                    # Final delta with stop_reason
                    stop_reason = event.get("delta", {}).get("stop_reason", "end_turn")
                    finish = _map_stop_reason(stop_reason)
                    usage = event.get("usage", {})
                    chunk = _build_stream_chunk(
                        completion_id, created, original_model, "",
                        finish_reason=finish,
                        usage=usage,
                    )
                    yield f"data: {json.dumps(chunk)}\n\n".encode("utf-8")

                elif event_type == "error":
                    err = event.get("error", {})
                    raise Exception(
                        f"Anthropic stream error: {err.get('type', 'unknown')}: "
                        f"{err.get('message', 'unknown error')}"
                    )

    yield b"data: [DONE]\n\n"


def _build_stream_chunk(
    completion_id: str,
    created: int,
    model: str,
    text: str,
    finish_reason: str | None = None,
    usage: dict[str, Any] | None = None,
    reasoning_content: str | None = None,
) -> dict[str, Any]:
    """Build a single OpenAI-format streaming chunk."""
    delta: dict[str, Any] = {}
    if text:
        delta["content"] = text
    if reasoning_content:
        delta["reasoning_content"] = reasoning_content

    chunk: dict[str, Any] = {
        "id": completion_id,
        "object": "chat.completion.chunk",
        "created": created,
        "model": model,
        "choices": [
            {
                "index": 0,
                "delta": delta,
                "finish_reason": finish_reason,
            }
        ],
    }
    if usage:
        chunk["usage"] = {
            "prompt_tokens": usage.get("input_tokens", 0),
            "completion_tokens": usage.get("output_tokens", 0),
            "total_tokens": usage.get("input_tokens", 0) + usage.get("output_tokens", 0),
        }
    return chunk


# ---------------------------------------------------------------------------
# Error helpers
# ---------------------------------------------------------------------------

def _wrap_error(exc: Exception) -> Exception:
    if isinstance(exc, httpx.HTTPStatusError):
        body = ""
        try:
            body = exc.response.text
        except Exception:  # noqa: BLE001
            pass
        return Exception(f"Anthropic API error {exc.response.status_code}: {body}")
    if isinstance(exc, httpx.ReadTimeout):
        return Exception("Anthropic API request timed out")
    return exc
